public/app/plugins/datasource/loki/formatterTS.ts

```diff
@@ -0,0 +1,406 @@
+import { SyntaxNode } from '@lezer/common';
+import { trimEnd } from 'lodash';
+
+import {
+  Eq,
+  Identifier,
+  String,
+  LogExpr,
+  Matcher,
+  parser,
+  PipelineExpr,
+  Selector,
+  Nre,
+  Re,
+  Neq,
+  LabelParser,
+  LabelFilter,
+  JsonExpressionParser,
+  LineFormatExpr,
+  LabelFormatExpr,
+  LineComment,
+  LineFilter,
+  Filter,
+  Regexp,
+  Pattern,
+  JsonExpression,
+  IpLabelFilter,
+  UnitFilter,
+  NumberFilter,
+  DurationFilter,
+  BytesFilter,
+  Gtr,
+  Gte,
+  Lss,
+  Lte,
+  Eql,
+  Duration,
+  Bytes,
+  Number,
+  LabelFormatMatcher,
+  FilterOp,
+} from '@grafana/lezer-logql';
+
+export const formatLogQL = (query: string): string => {
+  const tree = parser.parse(query);
+  let formatted = '';
+
+  tree.iterate({
+    enter: (ref): void => {
+      const node = ref.node;
+
+      switch (node.type.id) {
+        case Selector:
+          formatted += formatSelector(node, query);
+          break;
+
+        case PipelineExpr:
+          node.parent?.type.id === LogExpr && (formatted += formatPipelineExpr(node, query));
+          break;
+      }
+    },
+  });
+
+  return formatted;
+};
+
+export function formatSelector(node: SyntaxNode, query: string): string {
+  const selector = query.substring(node.from, node.to);
+  const subtree = parser.parse(selector);
+  const labelNodes: SyntaxNode[] = [];
+  let response = '';
+
+  // get all label nodes from the selector
+  subtree.iterate({
+    enter: (ref): void => {
+      const node = ref.node;
+      if (node.type.id === Matcher) {
+        labelNodes.push(node);
+      }
+    },
+  });
+
+  // sort the label nodes in alphabetical order
+  labelNodes.sort((a, b) => {
+    const LabelNodeA = a.getChild(Identifier)!;
+    const LabelNodeB = b.getChild(Identifier)!;
+
+    const labelValueA = query.substring(LabelNodeA.from, LabelNodeA.to);
+    const labelValueB = query.substring(LabelNodeB.from, LabelNodeB.to);
+
+    if (labelValueA < labelValueB) {
+      return -1;
+    }
+    if (labelValueA > labelValueB) {
+      return 1;
+    }
+    return 0;
+  });
+
+  // add the formatted label to the response
+  labelNodes.forEach((node) => {
+    const labelNode = node.getChild(Identifier)!;
+    const valueNode = node.getChild(String)!;
+    let operatorNode: SyntaxNode;
+
+    if (node.getChild(Eq)) {
+      operatorNode = node.getChild(Eq)!;
+    } else if (node.getChild(Neq)) {
+      operatorNode = node.getChild(Neq)!;
+    } else if (node.getChild(Re)) {
+      operatorNode = node.getChild(Re)!;
+    } else {
+      operatorNode = node.getChild(Nre)!;
+    }
```

Comment:
Since we know the query is valid from line 45, we can simplify this to just:

```suggestion
    operatorNode = labelNode.nextSibling;
```